### PR TITLE
IGNITE-11961  Provide JMX metrics for PME timings

### PR DESCRIPTION
**Pull request by NSAmelchev**
- add the isOperationsBlockedByPme metric

### CODE REVIEW

**Comment by anton-vinogradov** [comment]
checked the PR, LGTM. 
@nizhikov could you please perform final check?

**Comment by nizhikov** [comment]
@NSAmelchev 
@anton-vinogradov 

I thought we should **add** new metrics to calculate blocking period.
Don't think we should remove existing metrics for a whole PME duration.

In the end we should have:

* `CurrentPMEDuration` (existing metric, we shouldn't touch it).
* `CurrentPMECacheOperationsBlockedDuration` - new long metric. only in the new framework, don't appear in `ClusterMetrics` JMX bean.
* `PMEDuration` - new histogram.
* `PMECacheOperationsBlockedDuration` - new histogram.

WHYT?

**Comment by nizhikov** [comment]
`CurrentPMECacheOperationsBlockedDuration` - should be calculated similar to `CurrentPMEDuration`, but count only blocked duration

`PMECacheOperationsBlockedDuration` - should be updated on the finish of blocking PME phase.

`PMEDuration` - should be updated on the finish of PME.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/GridDhtTopologyFuture.java [hunk 1]
```diff
@@ -87,7 +87,6 @@
         @Nullable Collection<?> keys);
 
     /**
-     *
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/GridDhtTopologyFuture.java]: Why this change?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java [hunk 2]
```diff
@@ -2299,6 +2327,8 @@ private String exchangeTimingsLogMessage(String header, List<String> timings) {
                 if (err == null) {
                     timeBag.finishGlobalStage("Exchange done");
 
+                    updateDurationHistogram(TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTime));
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java]: I think millisecond precision is enough.
Let's change this to millis?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java [hunk 3]
```diff
@@ -2350,6 +2379,23 @@ private String exchangeTimingsLogMessage(String header, List<String> timings) {
...
+    /**
+     * Updates the {@link GridMetricManager#CACHE_OPERATIONS_BLOCKED_DURATION_HISTOGRAM} metric if needed.
+     *
+     * @param duration The total duration of the current PME.
+     */
+    private void updateDurationHistogram(long duration) {
+        if (isClientEventExchangeWithoutAffinityChange())
+            return;
+
+        MetricRegistry mreg = cctx.kernalContext().metric().registry(GridMetricManager.SYS_METRICS);
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java]: Can we store this metric in the class field instead of searching it on each PME?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/GridMetricManager.java [hunk 4]
```diff
@@ -135,6 +136,12 @@
     /** Daemon thread count metric name. */
     public static final String DAEMON_THREAD_CNT = "DaemonThreadCount";
 
+    /** Cache operations blocked duration metric name. */
+    public static final String CACHE_OPERATIONS_BLOCKED_DURATION = "cacheOperationsBlockedDuration";
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/metric/GridMetricManager.java]: The naming of the metrics differs. Why?
Let's make it consistent.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/GridMetricManager.java [hunk 5]
```diff
@@ -198,10 +205,16 @@ public GridMetricManager(GridKernalContext ctx) {
         sysreg.register(DAEMON_THREAD_CNT, threads::getDaemonThreadCount, null);
         sysreg.register("CurrentThreadCpuTime", threads::getCurrentThreadCpuTime, null);
         sysreg.register("CurrentThreadUserTime", threads::getCurrentThreadUserTime, null);
+
+        HistogramMetric cacheOperationsBlockedDurationHistogram =
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/metric/GridMetricManager.java]: Let's use `sysreg.histogram` instead of explicit instance creation.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/GridMetricManager.java [hunk 6]
```diff
@@ -495,6 +508,21 @@ private long totalSysMemory() {
         }
     }
 
+    /**
+     * @return Bounds for the {@link #CACHE_OPERATIONS_BLOCKED_DURATION_HISTOGRAM} metric. 30 bounds with 1-second
+     * interval by default.
+     */
+    private long[] getDurationHistogramBounds() {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/metric/GridMetricManager.java]: Why this method?
Let's write
`long[] bounds = new long[] {
0, 1000, 2000, 3000,... 30000
}`

These bounds look too much for me, also.
Can we start with `500, 1000, 5000, 30000` intervals?
User will be able to configure this.

**Comment by NSAmelchev** [comment]
> @NSAmelchev
> @anton-vinogradov
> 
> I thought we should **add** new metrics to calculate blocking period.
> Don't think we should remove existing metrics for a whole PME duration.
> 
> In the end we should have:
> 
>     * `CurrentPMEDuration` (existing metric, we shouldn't touch it, will be migrated to new framework in [IGNITE-11922](https://issues.apache.org/jira/browse/IGNITE-11922?jql=labels%20%3D%20IEP-35%20ORDER%20BY%20status%20ASC)).
> 
>     * `CurrentPMECacheOperationsBlockedDuration` - new long metric. only in the new framework, don't appear in `ClusterMetrics` JMX bean.
> 
>     * `PMEDuration` - new histogram.
> 
>     * `PMECacheOperationsBlockedDuration` - new histogram.
> 
> 
> WDYT?

@nizhikov Nikolay,  I have asked this on dev-list and Igniters says that we can update current existing metric(not released yet). It will become redundant with implementing new metrics.

I suggest implementing only two metrics as I have done.

**Comment by anton-vinogradov** [comment]
> @NSAmelchev
> @anton-vinogradov
> 
> I thought we should **add** new metrics to calculate blocking period.
> Don't think we should remove existing metrics for a whole PME duration.
> 
> In the end we should have:
> 
> * `CurrentPMEDuration` (existing metric, we shouldn't touch it, will be migrated to new framework in [IGNITE-11922](https://issues.apache.org/jira/browse/IGNITE-11922?jql=labels%20%3D%20IEP-35%20ORDER%20BY%20status%20ASC)).
> * `CurrentPMECacheOperationsBlockedDuration` - new long metric. only in the new framework, don't appear in `ClusterMetrics` JMX bean.
> * `PMEDuration` - new histogram.
> * `PMECacheOperationsBlockedDuration` - new histogram.
> 
> WDYT?

Mostly agree, but this should be performed at another issue.

**Comment by nizhikov** [comment]
> Mostly agree, but this should be performed at another issue.
> Let's start separate discussion for PME metrics improvements (necessary)

Why? Let's implement it here!

**Comment by nizhikov** [comment]
@NSAmelchev 

> I suggest implementing only two metrics as I have done.

Seems, we can go with the 4 metrics I propose.
It shouldn't be hard.

Please, implement it in this issue.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java [hunk 7]
```diff
@@ -476,6 +489,19 @@ else if (m instanceof GridDhtPartitionDemandLegacyMessage) {
...
+
+        MetricRegistry mreg = cctx.kernalContext().metric().registry(PME_METRICS);
+
+        mreg.register(CURRENT_PME_DURATION,
+            () -> getCurrentPMEDuration(false),
+            "Current PME duration in milliseconds.");
+
+        mreg.register(CURRENT_PME_CACHE_OPERATIONS_BLOCKED_DURATION,
+            () -> getCurrentPMEDuration(true),
+            "Current PME cache cperations blocked duration in milliseconds.");
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java]: typo in `cperations`

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java [hunk 8]
```diff
@@ -476,6 +489,19 @@ else if (m instanceof GridDhtPartitionDemandLegacyMessage) {
...
             }
         }
+
+        MetricRegistry mreg = cctx.kernalContext().metric().registry(PME_METRICS);
+
+        mreg.register(CURRENT_PME_DURATION,
+            () -> getCurrentPMEDuration(false),
+            "Current PME duration in milliseconds.");
+
+        mreg.register(CURRENT_PME_CACHE_OPERATIONS_BLOCKED_DURATION,
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java]: Lest rename to `PME_OPS_BLOCKED_DURATION`

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java [hunk 9]
```diff
@@ -476,6 +489,19 @@ else if (m instanceof GridDhtPartitionDemandLegacyMessage) {
                 });
             }
         }
+
+        MetricRegistry mreg = cctx.kernalContext().metric().registry(PME_METRICS);
+
+        mreg.register(CURRENT_PME_DURATION,
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java]: Let's rename to `PME_DURATION`

Comment (NSAmelchev) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java]: @nizhikov {{PME_DURATION}} is already exist. It's the histogram.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java [hunk 10]
```diff
@@ -2705,6 +2731,31 @@ else if (fut.isDone() && fut.topologyVersion().compareTo(from) >= 0)
...
+    private long getCurrentPMEDuration(boolean blocked) {
+        GridDhtPartitionsExchangeFuture fut = lastTopologyFuture();
+
+        return fut == null ? 0 : fut.getCurrentPMEDuration(blocked);
+    }
+
+    /**
+     * @return Histogram of PME durations metric.
+     */
+    public HistogramMetric getDurationHistogram() {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java]: According to the code style, we should use `durationHistogram` instead of `getDurationHistogram`

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java [hunk 11]
```diff
@@ -2705,6 +2731,31 @@ else if (fut.isDone() && fut.topologyVersion().compareTo(from) >= 0)
...
+        return durationHistogram;
+    }
+
+    /**
+     * @return Histogram of blocking PME durations metric.
+     */
+    public HistogramMetric getBlockingDurationHistogram() {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java]: According to the code style, we should use `blockingDurationHistogram` instead of `getBlockingDurationHistogram`

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java (hunk 8 above)
Comment (NSAmelchev) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java]: @nizhikov What do you think about `CURRENT_PME_OPS_BLOCKED_DURATION`? To avoid confusion with histograms.

File: modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/ClusterMetricsImpl.java [hunk 12]
```diff
@@ -163,6 +163,9 @@
      */
     private final IntMetric daemonThreadCnt;
 
+    /** Current PME duration in milliseconds. */
+    private final LongMetric currentPMEDuration;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/ClusterMetricsImpl.java]: `currentPMEDuration` -> `pmeDuration`

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/GridMetricManager.java [hunk 13]
```diff
@@ -198,10 +214,20 @@ public GridMetricManager(GridKernalContext ctx) {
...
+        pmeReg.histogram(PME_DURATION, pmeBounds,
+            "Histogram of PME durations in milliseconds.");
+
+        pmeReg.histogram(PME_CACHE_OPERATIONS_BLOCKED_DURATION, pmeBounds,
+            "Histogram of cache operations blocked PME durations in milliseconds.");
     }
 
     /** {@inheritDoc} */
-    @Override protected void onKernalStart0() throws IgniteCheckedException {
+    @Override protected void onKernalStart0() {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/metric/GridMetricManager.java]: Why this change?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java [hunk 14]
```diff
@@ -2705,6 +2731,31 @@ else if (fut.isDone() && fut.topologyVersion().compareTo(from) >= 0)
         return true;
     }
 
+    /**
+     * @param blocked {@code True} If take into account only cache operations blocked PME.
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java]: typo If -> if

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java [hunk 15]
```diff
@@ -2705,6 +2731,31 @@ else if (fut.isDone() && fut.topologyVersion().compareTo(from) >= 0)
...
+     * running PME or {@code blocked} was set to {@code true} and current PME don't block cache operations.
+     */
+    private long currentPMEDuration(boolean blocked) {
+        GridDhtPartitionsExchangeFuture fut = lastTopologyFuture();
+
+        return fut == null ? 0 : fut.currentPMEDuration(blocked);
+    }
+
+    /**
+     * @return Histogram of PME durations metric.
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java]: Can you make this javadoc one-liner.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java [hunk 16]
```diff
@@ -2705,6 +2731,31 @@ else if (fut.isDone() && fut.topologyVersion().compareTo(from) >= 0)
...
+     */
+    public HistogramMetric durationHistogram() {
+        return durationHistogram;
+    }
+
+    /**
+     * @return Histogram of blocking PME durations metric.
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java]: Can you make this javadoc one-liner.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java [hunk 17]
```diff
@@ -760,6 +777,8 @@ public void init(boolean newCrd) throws IgniteInterruptedCheckedException {
             else
                 state = cctx.kernalContext().clientNode() ? ExchangeLocalState.CLIENT : ExchangeLocalState.SRV;
 
+            initTime = System.currentTimeMillis();
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java]: Why this change?

File: modules/core/src/test/java/org/apache/ignite/spi/discovery/ClusterMetricsSelfTest.java [hunk 18]
```diff
@@ -0,0 +1,161 @@
...
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        cfg.setMetricsUpdateFrequency(METRIC_UPDATE_FREQUENCY);
+
+        cfg.setCommunicationSpi(new TestRecordingCommunicationSpi());
+
+        boolean client = igniteInstanceName.startsWith("client");
+
+        cfg.setClientMode(client);
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/spi/discovery/ClusterMetricsSelfTest.java]: Let's refactor this to `cfg.setClientMode(igniteInstanceName.startsWith("client"));`

File: modules/core/src/test/java/org/apache/ignite/spi/discovery/ClusterMetricsSelfTest.java [hunk 19]
```diff
@@ -0,0 +1,161 @@
...
+@GridCommonTest(group = "Utils")
+public class ClusterMetricsSelfTest extends GridCommonAbstractTest {
+    /** */
+    private static final long METRIC_UPDATE_FREQUENCY = 100;
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        cfg.setMetricsUpdateFrequency(METRIC_UPDATE_FREQUENCY);
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/spi/discovery/ClusterMetricsSelfTest.java]: Why do we need this frequency?

File: modules/core/src/test/java/org/apache/ignite/spi/discovery/ClusterMetricsSelfTest.java [hunk 20]
```diff
@@ -0,0 +1,159 @@
...
+
+        IgniteCache<Object, Object> cache = ignite.getOrCreateCache(
+            new CacheConfiguration<>(DEFAULT_CACHE_NAME)
+                .setAtomicityMode(TRANSACTIONAL));
+
+        cache.put(1, 1);
+
+        awaitPartitionMapExchange();
+
+        assertTrue(GridTestUtils.waitForCondition(() -> currentPMEDuration.value() == 0, 1000));
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/spi/discovery/ClusterMetricsSelfTest.java]: assertTrue -> assertEquals?

File: modules/core/src/test/java/org/apache/ignite/spi/discovery/ClusterMetricsSelfTest.java (hunk 19 above)
Comment (NSAmelchev) [reply modules/core/src/test/java/org/apache/ignite/spi/discovery/ClusterMetricsSelfTest.java]: For test speed optimization. Existing the `getCurrentPmeDuration` metric not covered by tests. I migrate it to the new framework.
I check it by a condition:` assertTrue(GridTestUtils.waitForCondition(() -> ignite.cluster().metrics().getCurrentPmeDuration() > 0, 1000));`
This means that the client will send metrics about itself and server node will get and check it.

File: modules/core/src/test/java/org/apache/ignite/spi/discovery/ClusterMetricsSelfTest.java [hunk 21]
```diff
@@ -0,0 +1,159 @@
...
+        assertTrue(currentBlockingPMEDuration.value() == 0);
+
+        // There was two blocking exchange: server node start and cache start.
+        assertTrue(Arrays.stream(durationHistogram.value()).sum() == 2);
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/spi/discovery/ClusterMetricsSelfTest.java]: assertTrue -> assertEquals?

File: modules/core/src/test/java/org/apache/ignite/spi/discovery/ClusterMetricsSelfTest.java (hunk 18 above)
Comment (NSAmelchev) [reply modules/core/src/test/java/org/apache/ignite/spi/discovery/ClusterMetricsSelfTest.java]: Done.

File: modules/core/src/test/java/org/apache/ignite/spi/discovery/ClusterMetricsSelfTest.java (hunk 21 above)
Comment (NSAmelchev) [reply modules/core/src/test/java/org/apache/ignite/spi/discovery/ClusterMetricsSelfTest.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java (hunk 16 above)
Comment (NSAmelchev) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java (hunk 15 above)
Comment (NSAmelchev) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java (hunk 14 above)
Comment (NSAmelchev) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java]: Thanks. Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/GridMetricManager.java (hunk 13 above)
Comment (NSAmelchev) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/metric/GridMetricManager.java]: Reverted.

File: modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/ClusterMetricsImpl.java (hunk 12 above)
Comment (NSAmelchev) [reply modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/ClusterMetricsImpl.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java (hunk 11 above)
Comment (NSAmelchev) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java]: Thanks. Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java (hunk 10 above)
Comment (NSAmelchev) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java]: Thanks. Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java (hunk 7 above)
Comment (NSAmelchev) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java]: Thanks. Done.
